File: V8-1/JT-Loader_Vis-Studio_C#/JTLoader/DataType/Int32CDP.cs
/////////////////////////////////////////////////////////////////////
//
// This class is used to store a Vector(List<>) collection of Int32's (pre version 9.5)
//
// Refer to 7.2.2.1.3 Tri-Strip Set Shape LOD Element page 124 - on (Ver 9.5 rev D)
//
/////////////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using System.Windows.Forms;
using C_sharp_JT_Reader.Codecs;

namespace C_sharp_JT_Reader
{
    public class Int32CDP
    {

        static byte[] _data;
        static int _filePosCount;
        static int[] result;
        static UInt32[] _codeTextWord;
        //static bool _probabiltyFirstPass;
        static List<string> _richTextBox;

        static Int32 _codeTextCount;
        static Int32[] _codeTextWords;

        static BitLengthCodec m_bitLengthCodec;
        static ArithmeticCodec m_arithmeticCodec;
        //Int32ProbabilityContexts _int32ProbabilityContexts;

        // Figure 169 : Int32 Compressed Data Packet data collection
        public static Int32[] Int32CDPone(CodecDriver.PredictorType value)
        {
//            _probabiltyFirstPass = true;
            return Int32CDPRead();
        }

        public static Int32[] Int32CDPRead()
        {
            Int32ProbabilityContexts _int32ProbabilityContexts = null;

            //_int32ProbabilityContexts = null;

            _richTextBox.Add("\n----------------------- Int32CDP data Collection ------------------------");

            byte[] fileBytes;
            Int32 _codeTextLength = 0;
            Int32 _valueElementCount = 0;

            byte codecType = _data[_filePosCount];
            string str8 = codecType.ToString();
            _richTextBox.Add("Codec Type = " + str8);
            _filePosCount += sizeof(byte);


            #region Huffman Codec
            // Huffman Codec
            if (codecType == 2)
            {
            }
            #endregion

            #region Arithmetic Codec
            // Arithmetic Codec
            if (codecType == 3)
            {
                fileBytes = new byte[4];
                // Int32 probabilityContextsMk2 Page 261
                //Figure 169 (Ver 8.1): Int32 Probability Contexts data collection

                _int32ProbabilityContexts = new Int32ProbabilityContexts(_richTextBox, _data, _filePosCount);

                // Retrieve the file position
                _filePosCount = _int32ProbabilityContexts.UpDateFilePos();


                // Get the number of Out Of Band Values
                fileBytes = new byte[4];
                Buffer.BlockCopy(_data, _filePosCount, fileBytes, 0, 4);
                Int32 _outOfBandValueCount = DataTypes.getInt32(fileBytes);
                _richTextBox.Add("Out Of Band Value Count = " + _outOfBandValueCount.ToString());
                _filePosCount += sizeof(Int32);



                if (_outOfBandValueCount > 0)
                {
                    // Int32 Compressed Data Packet -> Go back the start of this class..... (recursion see recursion ?)
                    int[] _oOOBValues = Int32CDPRead();

                    // Pass the 'Out of Band' values to the _int32ProbabilityContexts
                    _int32ProbabilityContexts.SetOOBValues(_oOOBValues);
                }
            }
            #endregion


            if (codecType != 0)
            {
                fileBytes = new byte[4];
                Buffer.BlockCopy(_data, _filePosCount, fileBytes, 0, 4);
                _codeTextLength = DataTypes.getInt32(fileBytes);
                _richTextBox.Add("Code Text Length = " + _codeTextLength.ToString());
                _filePosCount += sizeof(Int32);

                fileBytes = new byte[4];
                Buffer.BlockCopy(_data, _filePosCount, fileBytes, 0, 4);
                _valueElementCount = DataTypes.getInt32(fileBytes);
                _richTextBox.Add("Value Element Count = " + _valueElementCount.ToString());
                _filePosCount += sizeof(Int32);


                if (_int32ProbabilityContexts == null || _int32ProbabilityContexts.GetTableCount()==1)
                {
                    _codeTextCount = _valueElementCount;
                }
                else
                {
                    // VecI32 codeText
                    fileBytes = new byte[4];
                    Buffer.BlockCopy(_data, _filePosCount, fileBytes, 0, 4);
                    _codeTextCount = DataTypes.getInt32(fileBytes);
                    _richTextBox.Add("Code Text Count = " + _codeTextCount.ToString());
                    _filePosCount += sizeof(Int32);


                }
                _codeTextWords = new Int32[_codeTextLength];//(4 * _codeTextCount)];
                _codeTextWords = ReadCodeText(_codeTextCount, _codeTextWords);

            }


            result = new int[2];

            // Different Codecs required for decoding the codeText, etc.
            #region Codec Decoding
            if (codecType == 1)
            {
                // Bitlength Decoding Class Page 320
                m_bitLengthCodec = new BitLengthCodec(_richTextBox);
                // Activate the codec
                result = m_bitLengthCodec.DecompressBitLength(_codeTextWords, _codeTextLength, _valueElementCount);
            }

            // Arimetic Decoding 
            if (codecType == 3)
            {
                m_arithmeticCodec = new ArithmeticCodec(_richTextBox);
                // Activate the codec
                result = m_arithmeticCodec.DecodeArithmetic(_int32ProbabilityContexts, _codeTextWords, _codeTextLength, _codeTextCount, _valueElementCount);
            }
            #endregion


            _richTextBox.Add("-------------------- End of Int32CDP data Collection ---------------------");
            return result;
        }


        private static Int32[] ReadCodeText(int _codeTextCount, Int32[] _codeTextWords)
        {
            int _count = 0;
            byte[] _fileBytes = new byte[4];

            int conter = 0;
            Buffer.BlockCopy(_data, _filePosCount, _fileBytes, 0, 4);
            int counter = DataTypes.getInt32(_fileBytes);
            _filePosCount += sizeof(byte) * 4;


            int[] codeText = new int[counter * 4];

            //for (int j = 0; j < _codeTextCount; j++)
            for (int j = 0; j < counter; j++)
            {

                byte[] _reversedCodeText = new byte[4];

                // *** To-do *** //
                // if (littleEndian)

                // Code Text Word
                Buffer.BlockCopy(_data, _filePosCount, _fileBytes, 0, 4);
                _reversedCodeText = _fileBytes.Reverse().ToArray();
                _filePosCount += sizeof(byte) * 4;
                for (int i = 0; i < 4; i++)
                {
                    codeText[i + _count] = _reversedCodeText[i];
                    //_codeTextWords[i + _count] = _reversedCodeText[i];
                }
                _count += 4;

            }
            return codeText;// _codeTextWords;
        }

        public static void SetUpFilePosition(int filePosCount)
        {
            _filePosCount = filePosCount;
        }

        public static void SetUpData(byte[] data)
        {
            _data = data;
        }

        public static void SetupRTF(List<string> richTextBox)
        {
            _richTextBox = richTextBox;
        }

        public static int GetFilePos()
        {
            return _filePosCount;
        }

        public static int GetValueCount()
        {
            return 0;// _valueCount;
        }

        public static UInt32[] GetCodeTextWord()
        {
            return _codeTextWord;
        }
    }
}

File: V8-1/JT-Loader_Vis-Studio_C++/JTLoader/Int32CDP.cpp

#include "Int32CDP.h"

Int32CDP::Int32CDP(void)
{
}

Int32CDP::~Int32CDP(void)
{
}

Int32CDP::Int32CDP(INT32* filePointer, Reader reader)
{
	_filePosCount = filePointer;
	_reader = reader;
}

// Mthod to Traverse the codecs
vector<INT32> Int32CDP::traverseData(PredictorType ePredType)
{
	return Int32CDPRead();
}


vector<INT32> Int32CDP::Int32CDPRead()
{
	_contextSwitch = false;

	UCHAR codecType = _reader.getBtye(_filePosCount);

	#pragma region Huffman Codec
		if (codecType == 2)
		{
			MessageBox(0, TEXT("***** Not Implemented *****"), TEXT("Huffman Codec"), 0);
		}
	#pragma endregion

	#pragma region Arithmetic Codec
		if (codecType == 3)
		{
			_codecDriver1 = new CodecDriver();

            _int32ProbabilityContexts = new Int32ProbabilityContexts(_filePosCount, _reader);

			_contextSwitch = true;

            // Get the number of Out Of Band Values
            INT32 _outOfBandValueCount = _reader.getInt32(_filePosCount);

            if (_outOfBandValueCount > 0)
            {
				vector<int> _oOOBValues = Int32CDPRead();

                // Pass the 'Out of Band' values to the Probability Contexts
                _int32ProbabilityContexts->SetOOBValues(_oOOBValues);
            }
		}
#pragma endregion

#pragma region Null Codec
	if (codecType == 0)
	{
	}
#pragma endregion

	if (codecType != 0)
	{
		// The total number of bits of CodeText Data
		INT32 _codeTextLength = _reader.getInt32(_filePosCount);

		// The number of values that the Codec is expected to decode
		INT32 _valueElementCount = _reader.getInt32(_filePosCount);

		if (!_contextSwitch || _int32ProbabilityContexts->GetTableCount() == 1)
		{
			_codeTextCount = _valueElementCount;
			_contextSwitch = true;
		}
		else
		{
			// VecU32 codeText
			_codeTextCount = _reader.getInt32(_filePosCount);
		}

		// Count the number of codetext words required
		int counter = _reader.getInt32(_filePosCount);

		vector<UINT32> _codeTextWords;

		_codeTextWords.resize(counter * 4);

		for (int i = 0; i < counter * 4; i+=4)
		{
			ReadCodeText(_codeTextWords, i);
		}

#pragma region Bitlength Codec
		if (codecType == 1)
		{
			_bitLengthCodec = BitLengthCodec();
			_codecDriver1 = new CodecDriver(_codeTextWords,_codeTextLength, _valueElementCount);
			_bitLengthCodec.decode(_codecDriver1);
			_primitiveListIndices = _codecDriver1->getIndices();
		}
#pragma endregion

#pragma region Arimetic Codec
        if (codecType == 3)
        {	_arithmeticCodec = ArithmeticCodec();
			_codecDriver1 = new CodecDriver(_codeTextWords,_codeTextLength, _valueElementCount);
			_primitiveListIndices = _arithmeticCodec.DecodeArithmetic(_int32ProbabilityContexts, _codeTextWords, _codeTextLength, _codeTextCount, _valueElementCount);
        }
#pragma endregion

	}
	return _primitiveListIndices;
}

// Method to extract the CodeText Words & re-order if necessary
void Int32CDP::ReadCodeText(vector<UINT32> &_codeTextWords, INT32 _count)
{
    UCHAR* _fileBytes = new UCHAR[4];

	// Code Text Word
	_fileBytes[0] = _reader.getBtye(_filePosCount);
	_fileBytes[1] = _reader.getBtye(_filePosCount);
	_fileBytes[2] = _reader.getBtye(_filePosCount);
	_fileBytes[3] = _reader.getBtye(_filePosCount);

    // if littleEndian swap byte order
	if (_reader.getByteOrder() == 0)
	{
		_codeTextWords[0 + _count] = _fileBytes[3];
		_codeTextWords[1 + _count] = _fileBytes[2];
		_codeTextWords[2 + _count] = _fileBytes[1];
		_codeTextWords[3 + _count] = _fileBytes[0];
	}
	else
	{
		_codeTextWords[0 + _count] = _fileBytes[0];
		_codeTextWords[1 + _count] = _fileBytes[1];
		_codeTextWords[2 + _count] = _fileBytes[2];
		_codeTextWords[3 + _count] = _fileBytes[3];
	}
}